src/main/java/com/opentable/kafka/util/OffsetMetrics.java

```diff
@@ -0,0 +1,183 @@
+package com.opentable.kafka.util;
+
+import java.io.Closeable;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Map;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+import com.codahale.metrics.Metric;
+import com.codahale.metrics.MetricSet;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.opentable.concurrent.OTExecutors;
+import com.opentable.metrics.AtomicLongGauge;
+
+/**
+ * {@link MetricSet} implementation that monitors consumer groups consuming from topics.
+ *
+ * Runs a thread that polls the Kafka broker; use {@link #start()} and {@link #stop()} to spin up and down.
+ *
+ * The metric namespace is as follows.
+ *
+ * <p>
+ * {@code <topic>.{partition.%d,total}.{size,offset,lag}}
+ *
+ * <p>
+ * Size is the maximum offset of the topic (this is independent of any consumer). Offset is the current offset of the
+ * consumer. Lag is the size minus the offset. Totals are summed across the partitions.
+ */
+public class OffsetMetrics implements MetricSet, AutoCloseable, Closeable {
+    private static final Logger LOG = LoggerFactory.getLogger(OffsetMetrics.class);
+
+    private final String groupId;
+    private final Collection<String> topics;
+    private final Duration pollPeriod;
+    private final OffsetMonitor monitor;
+    private final Map<String, Metric> metricMap;
+    private final ScheduledExecutorService exec;
+
+    public OffsetMetrics(final String groupId, final String brokerList, final Collection<String> topics) {
+        this(groupId, brokerList, topics, Duration.ofSeconds(10));
+    }
+
+    @VisibleForTesting
+    OffsetMetrics(
+            final String groupId,
+            final String brokerList,
+            final Collection<String> topics,
+            final Duration pollPeriod) {
+        Preconditions.checkArgument(!topics.isEmpty(), "no topics");
+        this.groupId = groupId;
+        this.topics = topics;
+        this.pollPeriod = pollPeriod;
+        monitor = new OffsetMonitor(groupId, brokerList);
+
+        final Collection<String> badTopics = new ArrayList<>();
+        final ImmutableMap.Builder<String, Metric> builder = ImmutableMap.builder();
+        topics.forEach(topic -> {
+            final Map<Integer, Long> sizes = monitor.getTopicSizes(topic);
+            if (sizes.isEmpty()) {
+                badTopics.add(topic);
+                return;
+            }
+            for (final int part : sizes.keySet()) {
+                builder.put(String.format("%s.partition.%d.size",   topic, part), new AtomicLongGauge());
```

Comment:
Consider factoring these out e.g. `metricName(topic, part, SIZE)`

Reply:
Great suggestion.  In fact, I did exactly this in my work to have the `OffsetMetrics` take a `MetricRegistry`.  Update forthcoming.

Reply:
See 334fbcff4249b1442a41237e75e9f26e3bd03ff7.